src/global/project/environment.rs

```diff
@@ -63,3 +80,123 @@ pub struct ParseEnvironmentNameError {
     /// The string that was attempted to be parsed.
     pub attempted_parse: String,
 }
+
+/// Figures out what the status is of the exposed binaries of the environment.
+///
+/// Returns a tuple of the exposed binaries to remove and the exposed binaries to add.
+pub(crate) async fn get_expose_scripts_sync_status(
+    bin_dir: &BinDir,
+    env_dir: &EnvDir,
+    exposed: &IndexMap<ExposedName, String>,
+) -> miette::Result<(IndexSet<PathBuf>, IndexSet<ExposedName>)> {
+    // Get all paths to the binaries from the scripts in the bin directory.
+    let locally_exposed = bin_dir.files().await?;
+    let executable_paths = futures::future::join_all(locally_exposed.iter().map(|path| {
+        let path = path.clone();
+        async move {
+            extract_executable_from_script(&path)
+                .await
+                .ok()
+                .map(|exec| (path, exec))
+        }
+    }))
+    .await
+    .into_iter()
+    .flatten()
+    .collect::<Vec<_>>();
+
+    // Filter out all binaries that are related to the environment
+    let related_exposed = executable_paths
+        .into_iter()
+        .filter(|(_, exec)| exec.starts_with(env_dir.path()))
+        .map(|(path, _)| path)
+        .collect_vec();
+
+    // Get all related expose scripts not required by the environment manifest
+    let to_remove = related_exposed
+        .iter()
+        .filter(|path| {
+            !exposed
+                .iter()
+                .any(|(exposed_name, _)| executable_from_path(path) == exposed_name.to_string())
+        })
+        .cloned()
+        .collect::<IndexSet<PathBuf>>();
+
+    // Get all required exposed binaries that are not yet exposed
+    let mut to_add = IndexSet::new();
+    for (exposed_name, _) in exposed.iter() {
+        if related_exposed
+            .iter()
+            .map(|path| executable_from_path(path))
+            .any(|exec| exec == exposed_name.to_string())
+        {
+            to_add.insert(exposed_name.clone());
+        }
+    }
```

Comment:
Would be nice to make this more functional

Reply:
Done